src/shim.c

```diff
@@ -151,100 +155,224 @@ err_exit(const char *format, ...)
 	exit(EXIT_FAILURE);
 }
 
-/*!
- * Construct message in the proxy ctl rpc protocol format
- * Proxy control message format:
- *
- * | length  | Reserved| Data(Request/Response   |
- * | . . . . | . . . . | . . . . . . . . . . . . |
- * 0         4         8                         length
- *
- * \param json Json Payload
- * \param[out] len Length of the message
+/*
+ * Prepare byte stream in the proxy protocol format.
  *
- * \return Newly allocated string on sucess, NULL on failure
+ * \param header \ref frame_header
+ * \param payload Payload to be sent
+ * \param len Total size of the header and payload
  */
-char*
-get_proxy_ctl_msg(const char *json, size_t *len) {
-	char   *proxy_ctl_msg = NULL;
-	size_t  json_len;
+uint8_t*
+serialize_message(struct frame_header *header, const uint8_t *payload, ssize_t len)
+{
+	uint8_t *msg = NULL;
 
-	if (! (json && len)) {
+	if (! (header && len)) {
 		return NULL;
 	}
 
-	json_len = strlen(json);
-	*len = json_len + PROXY_CTL_HEADER_SIZE;
-	proxy_ctl_msg = calloc(*len + 1, sizeof(char));
+	msg = calloc((size_t)len, sizeof(uint8_t));
 
-	if (! proxy_ctl_msg) {
+	if (! msg) {
```

Comment:
There are callers of this function which check the return value, but the value will always be true due to this check. Should it return `NULL` rather than calling `abort()` maybe?